pkg/machines/app.jsx

```diff
@@ -18,21 +18,71 @@
  */
 import React from 'react';
 import PropTypes from 'prop-types';
+import { ToastNotificationList, ToastNotification } from 'patternfly-react';
 
 import HostVmsList from "./hostvmslist.jsx";
 import { StoragePoolList } from "./components/storagePools/storagePoolList.jsx";
 import { NetworkList } from "./components/networks/networkList.jsx";
 import LibvirtSlate from "./components/libvirtSlate.jsx";
 import { CreateVmAction } from "./components/create-vm-dialog/createVmDialog.jsx";
 import { AggregateStatusCards } from "./components/aggregateStatusCards.jsx";
+import { InlineNotification } from './components/notification/inlineNotification.jsx';
 
 class App extends React.Component {
     constructor(props) {
         super(props);
         this.state = {
-            activeTab: 1
+            activeTab: 1,
+            notifications: [],
+            /* Dictionary with keys being a resource's UUID and values the number of active error notifications for that resource */
+            resourceHasError: {},
         };
         this.changeActiveList = this.changeActiveList.bind(this);
+        this.onAddErrorNotification = this.onAddErrorNotification.bind(this);
+        this.onDismissErrorNotification = this.onDismissErrorNotification.bind(this);
+    }
+
+    /*
+     * Adds a new notification object to the notificationsArray. It also updates
+     * the error count for a specific resource.
+     * @param {object} notification - The notification object to be added to the array.
+     */
+    onAddErrorNotification(notification) {
+        let resourceHasError = Object.assign({}, this.state.resourceHasError);
+
+        if (resourceHasError[notification.resourceId])
+            resourceHasError[notification.resourceId]++;
+        else
+            resourceHasError[notification.resourceId] = 1;
+
+        notification.index = this.state.notifications.length;
```

Comment:
This is prone to conflicts. Assume you have one notification already, and add a second one, then `.index == 1`. Now delete the first notification, add a new one, its index will once again be `1`. Dismissing that notification will get rid of the older "1", but it's not clear which one the user actually clicked on.

Instead I recommend maintaining a `notificationId` counter in the state which only ever gets increased here, and not look at the array length at all. This avoids the conflict.

Then it's probably also better to make `notifications` a dictionary instead of an array, as otherwise over time you'll get bigger and bigger holes in it.

Reply:
That was horrible buggy code, thanks for noticing this.
I also have the exact same code in cockpit-podman, I will have to fix that as well.